subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskSuccessResultPostProcessor.java

```diff
@@ -37,43 +39,35 @@
-        if (buildOperation.getDetails() instanceof ExecuteTaskBuildOperationType.Details) {
-            parentsOfOperationsWithJavaCompileTaskAncestor.put(buildOperation.getId(), TASK_MARKER);
-        } else if (buildOperation.getParentId() != null && parentsOfOperationsWithJavaCompileTaskAncestor.containsKey(buildOperation.getParentId())) {
-            parentsOfOperationsWithJavaCompileTaskAncestor.put(buildOperation.getId(), buildOperation.getParentId());
-        }
     }
 
     @Override
     public void finished(BuildOperationDescriptor buildOperation, OperationFinishEvent finishEvent) {
         if (finishEvent.getResult() instanceof CompileJavaBuildOperationType.Result) {
             CompileJavaBuildOperationType.Result result = (CompileJavaBuildOperationType.Result) finishEvent.getResult();
-            OperationIdentifier taskBuildOperationId = findTaskOperationId(buildOperation.getParentId());
-            results.put(taskBuildOperationId, result);
-        }
-        parentsOfOperationsWithJavaCompileTaskAncestor.remove(buildOperation.getId());
-    }
-
-    private OperationIdentifier findTaskOperationId(OperationIdentifier id) {
-        Object parent = parentsOfOperationsWithJavaCompileTaskAncestor.get(id);
-        if (parent == TASK_MARKER) {
-            return id;
+            CompileJavaBuildOperationType.Details details = (CompileJavaBuildOperationType.Details) buildOperation.getDetails();
+            if (details == null) {
+                throw new IllegalStateException("No details for " + buildOperation.getDisplayName() + ", which is required for proper result tracking");
+            }
+            results.put(details.getTaskIdentityPath(), result);
         }
-        return findTaskOperationId((OperationIdentifier) parent);
     }
 
     @Override
-    public AbstractTaskResult process(AbstractTaskResult taskResult, OperationIdentifier taskBuildOperationId) {
-        CompileJavaBuildOperationType.Result compileResult = results.remove(taskBuildOperationId);
-        if (taskResult instanceof DefaultTaskSuccessResult && compileResult != null) {
-            return new DefaultJavaCompileTaskSuccessResult((DefaultTaskSuccessResult) taskResult, toAnnotationProcessorResults(compileResult.getAnnotationProcessorDetails()));
+    public AbstractTaskResult process(AbstractTaskResult taskResult, TaskInternal taskInternal) {
+        CompileJavaBuildOperationType.Result compileResult = results.remove(taskInternal.getIdentityPath().getPath());
+        if (taskResult instanceof DefaultTaskSuccessResult) {
+            if (compileResult != null) {
+                return new DefaultJavaCompileTaskSuccessResult((DefaultTaskSuccessResult) taskResult, toAnnotationProcessorResults(compileResult.getAnnotationProcessorDetails()));
+            } else if (taskInternal instanceof JavaCompile) {
+                LOGGER.warn("No compile result for " + taskInternal.getIdentityPath());
```

Comment:
I think we should make this `info` so we can see it if we suspect there's a problem, but otherwise, it's ignorable.

Reply:
Wow, it was `info` at some point. I guess copilot must have written it to `warn` when I was refactoring :(